homeassistant_api/rawbaseclient.py

```diff
@@ -73,20 +79,32 @@ def prepare_get_entity_histories_params(
         end_timestamp: Optional[datetime] = None,
         significant_changes_only: bool = False,
     ) -> Tuple[Dict[str, Optional[str]], str]:
-        """Pre-logic for `Client.get_entity_histories` and `Client.async_get_entity_histories`."""
+        """
+        Pre-logic for :py:meth:`Client.get_entity_histories` and :py:meth:`Client.async_get_entity_histories`.
+
+        Ensure timestamps
+
+        * use second resolution (microseconds are truncated)
+        * are timezone-aware
+        * are URL-encoded (as :py:meth:`construct_params` is used instead of request's default parameter encoding)
+        """
         params: Dict[str, Optional[str]] = {}
         if entities is not None:
             params["filter_entity_id"] = ",".join([ent.entity_id for ent in entities])
-        if end_timestamp is not None:
-            params["end_time"] = (
-                end_timestamp.isoformat()
-            )  # Params are automatically URL encoded
-        if significant_changes_only:
-            params["significant_changes_only"] = None
         if start_timestamp is not None:
+            start_timestamp = start_timestamp.replace(microsecond=0)
+            if start_timestamp.tzinfo is None:
+                start_timestamp = start_timestamp.astimezone()
             url = join("history/period/", start_timestamp.isoformat())
         else:
             url = "history/period"
+        if end_timestamp is not None:
+            end_timestamp = end_timestamp.replace(microsecond=0) + timedelta(seconds=1)
+            if end_timestamp.tzinfo is None:
+                end_timestamp = end_timestamp.astimezone()
+            params["end_time"] = end_timestamp.isoformat()
```

Comment:
Fun fact: We actually don't need to url-encode the parameters because requests and aiohttp actually urlencode it already. We format it as a string, they urlencode that string. The error i was running into when testing was that the colons in the timestamps were getting double-encoded essentially (colon would be urlencoded by us, then requests lets say, would urlencode the percent char in the urlencoded version of the colon).

Reply:
Nevermind, we need to urlencode the timestamps, but requests and aiohttp were encoding the query string a second time, despite it already being encoded.